### PR TITLE
Build infinite scroll experience

### PR DESCRIPTION
**Pull request by thomasthesecond**

### CODE REVIEW

**Comment by jcreamer898** [comment]
O_O

File: src/components/article/article.js [hunk 1]
```diff
@@ -0,0 +1,616 @@
+/* global googletag, utag */
+import { Component } from "../../core/bane";
+import $ from "jquery";
+import debounce from "lodash/function/debounce";
+import ArticleBodyComponent from "../article_body";
+import track from "../../core/decorators/track";
+
+export default class ArticleComponent extends Component {
+  initialize() {
+    this.canUseScrollFeature = (window.history && window.history.pushState) ? true : false;
```
Comment (jcreamer898) [inline src/components/article/article.js]: Don't need a ternary here... just,

```js
this.canUseScrollFeature = window.history && window.history.pushState;
```

File: src/components/article/article.js [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+    this.listOfArticles = [];
+    this.paginatedArticles = {};
+
+    this.delay = 1500;
+
+    this.nextSlotId = 1;
+
+    this._slugifyPlaceData();
+
+    window.lp.ads = {
```
Comment (jcreamer898) [inline src/components/article/article.js]: I kinda feel like this should go over in `lp_js_details.hbs`...

File: src/components/article/article.js [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+    this.howManyArticlesHaveLoaded = 1;
+    this.$activeArticle = this.$el.addClass("is-active");
+
+    let article = window.lp.article;
+
+    // Add the first article to the map
+    this.articles.set(this.$el[0], {
+      title: this.$activeArticle.data("title"),
+      slug: this.$activeArticle.data("slug"),
+      tealium: {
```
Comment (jcreamer898) [inline src/components/article/article.js]: Let's not use the word `tealium`, call it like `dimensions` or something...?

File: src/components/article/article.js [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+    // setTimeout is because of webfonts and images; probably need to use
+    // hasImagesLoaded and a webfont loader
+    setTimeout(() => {
+      this.articleHeight = this.$el.height();
+      this.articlePaginationHeight = this.$el.find(".article-pagination").outerHeight(true);
+      this.articleOffsetBottom = $(document).height() - this.articleHeight - this.articleOffsetTop;
+      this.scrollOffset = this.articleOffsetBottom + this.articleOffsetTop + this.articlePaginationHeight;
+      this.articleScrollTop = this.articleHeight - this.scrollOffset;
+      this.heightOfAllArticles += this.articleScrollTop;
+    }, this.delay);
```
Comment (jcreamer898) [inline src/components/article/article.js]: Have you tested the delay to see if we can shorten it at all? `1500` seems really long...

File: src/components/article/article.js [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+      });
+    });
+  }
+
+  /**
+   * Creates the initial list of articles and sets the next article that will be
+   * loaded; this method is to be called when the component initializes
+   */
+  _createInitialListOfArticles() {
+    this._getRelatedArticles(`${window.location.pathname}.json`).then((response) => {
```
Comment (jcreamer898) [inline src/components/article/article.js]: Can you pull the anonymous function into it's own method?

File: src/components/article/article.js [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+   */
+  _setNextArticle() {
+    this.nextArticle = this.listOfArticles[this.howManyArticlesHaveLoaded - 1];
+  }
+
+  /**
+   * Runs methods when scrolling
+   */
+  _scrollToNextArticle() {
+    console.log("Ready");
```
Comment (jcreamer898) [inline src/components/article/article.js]: Remove `console.log`

File: src/components/article/article.js [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+
+    return exists;
+  }
+
+  /**
+   * Use an AJAX call to get data for a new article
+   * @param {String} slug Pathname of article to get
+   */
+  _getNextArticle(slug) {
+    console.log(`Article ${this.howManyArticlesHaveLoaded + 1} is loading…`);
```
Comment (jcreamer898) [inline src/components/article/article.js]: Remove `console.log`

File: src/components/article/article.js [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+
+    this.isNextArticleLoading = true;
+
+    $(this.loader({}))
+      .appendTo(this.$activeArticle);
+
+    $.ajax(slug, {
+      success: (data) => {
+        this.$newArticle = $(this.template(data))
```
Comment (jcreamer898) [inline src/components/article/article.js]: Make the success point to a method, try not to use anonymous functions in ajax callbacks.

File: src/components/article/article.js [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+        this._addNewArticlesToArray(data.article.related_articles.articles);
+        this._updateNewArticle();
+        this._resetArticleDimensions();
+
+        this.isNextArticleLoading = false;
+
+        $(".article-loading").remove();
+
+        console.log(`Article ${this.howManyArticlesHaveLoaded} is done`);
+      },
```
Comment (jcreamer898) [inline src/components/article/article.js]: `console.log()`

File: src/components/article/article.js [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+      error: () => {
+        this.isNextArticleLoading = false;
+
+        // @TODO Maybe do something more substantial here
+        console.log(`Article ${this.howManyArticlesHaveLoaded + 1} is done with errors`);
```
Comment (jcreamer898) [inline src/components/article/article.js]: `console.log`

File: src/components/article/article.js (hunk 3 above)
Comment (thomasthesecond) [inline src/components/article/article.js]: This is how we're calling it in the data payload from the back-end, so I wanted to match the naming. I think we'd want to update the back-end as well.

File: src/components/article_body/article_body.js [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+ */
+export default class ArticleBodyComponent extends Component {
+  initialize() {
+    this.imageContainerSelector = ".stack__article__image-container";
+
+    this.loadImages().then(() => {
+      this.gallery = new ImageGallery({ el: this.$el });
+    });
+
+    // matchMedia("(min-width: 1200px)", (query) => {
```
Comment (jcreamer898) [inline src/components/article_body/article_body.js]: No need to leave commented out code.

File: src/components/article/article.js (hunk 4 above)
Comment (thomasthesecond) [inline src/components/article/article.js]: I haven't, but I think we can shorten it. My local env is usually slower.

File: src/components/article/article.js (hunk 3 above)
Comment (jcreamer898) [inline src/components/article/article.js]: Fair enough.

**Pull request merged**